Fix crash when issuing MMU commands and LCD corruption

gudnimg opened the pull request:
* a6a8996abe3338396adbf4c03d4db71117150745 Fix crash in `manage_response`
The watchdog needs to be kicked in the `for(;;)` loop.
This is usually done by calling `manage_heater` before
`manage_inactivity`.

Fixes for ProgressCode messages:
* d0a6ac24eee850c2dcb3f0a75f02f6f84c6a2ecd  Fix PROGMEM read in ProgressCode message
a garbled string and the same corruption would appear on the Status
Screen's status line.
* 05823d8586aeb1c6697c0fedc85ecd43bf9a3274 Fix overwrite in ProgressCode message buffer
When `snprintf()` if called we need to read the return value to see how
many bytes were written. Then when we call `strncpy_P()` through
`TranslateProgress()`, we need to tell the code to start writing
at byte '`len`', or `&msg[len]`. Also we need to update the byte size
which `strncpy_P()` is allowed to write (64 - len).
* e93bcca88205ac811df388fed9b894d1a4a97849 Add missing space in ProgressCode message
Such that:
`MMU2:P=123EngageIdler`
Becomes:
`MMU2:P=123 EngageIdler`

Fixes for ErrorCode messages:
* 13120d5bfb0e99bfea3cfa93eeeb462cc780da09 Fix PROGMEM read in ErrorCode message
This could cause the printer to crash. The Serial Stream would show
a garbled string.
Also included in this commit:
    * The buffer size is increased from 64 bytes to 192 bytes.  We need to
take into account the length of the ErrorTitle and ErrorDescription.
    * Fix overwrite in ErrorCode message buffer.
* cf4e3142f9fbd0a36f475c64af6d57d3de4fbcef Add missing spaces to ErrorCode messages
Such that:
`MMU2:E=32766ErrorTitleTextDescription`
Becomes:
`MMU2:E=32766 ErrorTitle TextDescription`
Also simplified the process of combining ErrorTitle and TextDescription
into the `msg` buffer by using `snprintf`. This is saver since we only
use `dstSize` in one place instead of two.

Image of Error Code after changes:
![image](https://user-images.githubusercontent.com/8218499/164969922-443398b3-a097-4d25-8551-f4699f6d4b51.png)

Image of Progress Code after changes:
![image](https://user-images.githubusercontent.com/8218499/164970032-0133be1e-9c53-4dea-bf1d-983081b96b25.png)

leptun commented on Firmware/mmu2.cpp:
```diff
@@ -511,6 +512,9 @@ void MMU2::manage_response(const bool move_axes, const bool turn_off_nozzle) {
         // - failed -> then do the safety moves on the printer like before
         // - finished ok -> proceed with reading other commands
+#ifdef WATCHDOG
+        wdt_reset();
```
Manually setting the watchdog is never the solution. It means that the heaters are also not updated. The `manage_inactivity(true)` lower is also suspicious since that one calls the `MMU2::mmu2.mmu_loop();` internally

gudnimg replied:
Indeed I noticed it is that way also on current MK3 branch so I didn’t touch it.

gudnimg replied:
Perhaps we should be calling `manage_heater()` just before `manage_inactivity()`. It is done in most places in the firmware and it kick the watchdog for us. If I search for `manage_inactivity()` then they are usually used together like this:

```c
manage_heater();
manage_inactivity();
```

gudnimg replied:
I am getting random axis crashes after this change. This looks strange:
![image](https://user-images.githubusercontent.com/8218499/164912907-8904f5a0-0ba8-4af5-8327-fa59764529eb.png)

gudnimg replied:
I didn't see the axis problem again. The E32903 seems to be an error code, I fixed it by pushing the middle button (I think?) on the MMU and that made the unit re-home.

gudnimg committed:
Fix crash in manage_response

The watchdog needs to be kicked in the for(;;) loop.
This is usually done by calling manage_heater before
manage_inactivity.

gudnimg committed:
Fix PROGMEM read in ProgressCode message

This could cause the printer to crash. The Serial Stream would show
a garbled string and the same corruption would appear on the Status
Screen's status line.

gudnimg committed:
Fix overwrite in ProgressCode message buffer

When snprintf() if called we need to read the return value to see how
many bytes were written. Then when we call strncpy_P() through
TranslateProgress(), we need to tell the code to start writing
at byte 'len', or &msg[len]. Also we need to update the byte size
which strncpy_P() is allowed to write (64 - len).

gudnimg commented on Firmware/mmu2_progress_converter.cpp:
```diff
@@ -62,7 +62,8 @@ static const char * const progressTexts[] PROGMEM = {
-    return ( pc <= 26 ) ? progressTexts[pc] : progressTexts[0]; // @@TODO ?? a better fallback option?
+    // @@TODO ?? a better fallback option?
+    return ( pc <= 26 ) ? static_cast<const char * const>(pgm_read_ptr(&progressTexts[pc])) : static_cast<const char * const>(pgm_read_ptr(&progressTexts[0]));
 }
```
I saw this solution from `first_lay_cal.cpp` where we are doing a similar thing of reading a list of `PROGMEM `strings.

DRracer replied:
so that's the way of forcing the compiler to understand, that the pointer point into PROGMEM :+1:

gudnimg committed:
Fix PROGMEM read in ErrorCode message

This could cause the printer to crash. The Serial Stream would show
a garbled string.

Also included in this commit:
* The buffer size is increased from 64 bytes to 192 bytes.  We need to
take into account the length of the ErrorTitle and ErrorDescription.
* Fix overwrite in ErrorCode message buffer.

gudnimg committed:
Add missing space in ProgressCode message

Such that:
MMU2:P=123EngageIdler
Becomes:
MMU2:P=123 EngageIdler

gudnimg committed:
Add missing spaces to ErrorCode messages

Such that:
MMU2:E=32766ErrorTitleTextDescription
Becomes:
MMU2:E=32766 ErrorTitle TextDescription

Also simplified the process of combining ErrorTitle and TextDescription
into the msg buffer by using snprintf. This is saver since we only
use dstSize in one place instead of two.

gudnimg commented:
I've updated the commit messages and PR description to help with review. I did not try to optimize the changes, only fix the crashes 😄

DRracer commented on Firmware/mmu2.cpp:
```diff
@@ -511,7 +511,7 @@ void MMU2::manage_response(const bool move_axes, const bool turn_off_nozzle) {
         // - finished ok -> proceed with reading other commands
-        
+        manage_heater();
         // @@TODO this needs verification - we need something which matches Marlin2's idle()
```
this @@TODO can be removed if `manage_heater()` and `manage_inactivity()` do their job

DRracer commented on Firmware/mmu2.cpp:
```diff
@@ -608,19 +608,23 @@ void MMU2::ReportError(ErrorCode ec) {
-        snprintf(msg, sizeof(msg), "MMU2:E=%hu", (uint16_t)ec);
+        // Log error format: MMU2:E=32766 ErrorTitle TextDescription
+
+        // The longest error description in errors_list.h is 144 bytes.
```
this is subject for removal - it was kept here just to prevent the compiler from stripping the texts completely.

DRracer commented on Firmware/mmu2.cpp:
```diff
@@ -633,13 +637,13 @@ void MMU2::ReportError(ErrorCode ec) {
     // Log progress - example: MMU2:P=123 EngageIdler
     char msg[64];
-    snprintf(msg, sizeof(msg), "MMU2:P=%hu", (uint16_t)pc);
+    int len = snprintf(msg, sizeof(msg), "MMU2:P=%hu ", (uint16_t)pc);
```
also subject for a change - no need to format it into a RAM buffer when the texts are in PROGMEM

DRracer commented:
@gudnimg well done making the code actually run :wink:
